Account Service Optimisation

Pubmatic-Dhruv-Sonone opened the pull request:
Related Issue: https://github.com/prebid/prebid-server/issues/2429

bsardo commented:
@Pubmatic-Dhruv-Sonone it looks like you'll need to merge with master to resolve the format issues you're seeing during the go 1.19 validation step.

Pubmatic-Dhruv-Sonone commented on account/account.go:
```diff
@@ -47,39 +48,6 @@ func GetAccount(ctx context.Context, cfg *config.Configuration, fetcher stored_r
-		if err == nil {
-			err = json.Unmarshal(completeJSON, account)
-
-			// this logic exists for backwards compatibility. If the initial unmarshal fails above, we attempt to
-			// resolve it by converting the GDPR enforce purpose fields and then attempting an unmarshal again before
-			// declaring a malformed account error.
-			// unmarshal fetched account to determine if it is well-formed
-			if _, ok := err.(*json.UnmarshalTypeError); ok {
-				// attempt to convert deprecated GDPR enforce purpose fields to resolve issue
-				completeJSON, err = ConvertGDPREnforcePurposeFields(completeJSON)
-				// unmarshal again to check if unmarshal error still exists after GDPR field conversion
-				err = json.Unmarshal(completeJSON, account)
-
-				if _, ok := err.(*json.UnmarshalTypeError); ok {
-					return nil, []error{&errortypes.MalformedAcct{
-						Message: fmt.Sprintf("The prebid-server account config for account id \"%s\" is malformed. Please reach out to the prebid server host.", accountID),
-					}}
-				}
-			}
-		}
```
Missed this during master merge. Added this logic.

Pubmatic-Dhruv-Sonone commented:
@bsardo  Can you please review the PR? I have addressed above comments.

mansinahar commented on stored_requests/fetcher.go:
```diff
@@ -28,7 +29,7 @@ type Fetcher interface {
 type AccountFetcher interface {
 	// FetchAccount fetches the host account configuration for a publisher
-	FetchAccount(ctx context.Context, accountID string) (json.RawMessage, []error)
+	FetchAccount(ctx context.Context, accountDefaultJSON json.RawMessage, accountID string) (*config.Account, []error)
```
I think we should leave the return type for `FetchAccount` to be a raw JSON string as our cache is currently configured to store stuff as JSON strings rather than objects. This is something we want ideally but IMHO it doesn't make sense until we change the cache code to accept objects as well.

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -163,7 +167,39 @@ func (fetcher *HttpFetcher) FetchAccount(ctx context.Context, accountID string)
-	return accountJSON, nil
+	// accountID resolved to a valid account, merge with AccountDefaults for a complete config
+	account := &config.Account{}
+	completeJSON, err := jsonpatch.MergePatch(accountDefaultsJSON, accountJSON)
```
With leaving the return type as raw JSON for `FetchAccount`, you can just return after this step and move the rest of the GDPR related backwards compatibility logic in the original `GetAccount` function itself. This will minimize the duplicate code needed in each of the fetcher implementation.

Pubmatic-Dhruv-Sonone replied:
Done

mansinahar commented on stored_requests/fetcher.go:
```diff
@@ -210,18 +211,27 @@ func (f *fetcherWithCache) FetchResponses(ctx context.Context, ids []string) (da
+	account, errs = f.fetcher.FetchAccount(ctx, accountDefaultJSON, accountID)
 	if len(errs) == 0 {
-		f.cache.Accounts.Save(ctx, map[string]json.RawMessage{accountID: account})
+		accountJSON, err := json.Marshal(account)
```
Once you make those above changes, you won't have to do JSON marshalling here before saving it in the cache as you'll already have it as raw JSON string from `FetchAccount`

Pubmatic-Dhruv-Sonone replied:
Reverted the account marshalling changes.

mansinahar reviewed (commented):
@Pubmatic-Dhruv-Sonone Thanks for taking the effort to make Prebid Server more efficient. I like the motivation behind this. However, given that we do have fetchers refactor on our radar some time for next year, I think we can keep the changes minimal while achieving what you're aiming for. Please see my comments below.

bsardo commented:
@Pubmatic-Dhruv-Sonone just curious what your thoughts are regarding the comments we left above.

Pubmatic-Dhruv-Sonone commented:
> @Pubmatic-Dhruv-Sonone just curious what your thoughts are regarding the comments we left above.

@bsardo Sorry for the delay, I was on leave last few weeks. I have addressed the above comments.

mansinahar commented on stored_requests/backends/http_fetcher/fetcher_test.go:
```diff
@@ -249,7 +249,7 @@ func newAccountHandler(t *testing.T, expectAccIDs []string, jsonifier func(strin
 		for _, accID := range gotAccIDs {
 			if accID != "" {
-				accIDResponse[accID] = jsonifier(accID)
+				accIDResponse[accID] = json.RawMessage(`{"id":"` + accID + `"}`)
```
Why not just use the existing `jsonifier` function that's been sent as parameter to `newAccountHandler`?

Pubmatic-Dhruv-Sonone replied:
We need a valid account json for this test case, the existing value was fails in json patching.  Removed jsonifier as I am already creating a valid json here and jsonifier adds escape characters to json string, which causes unmarshalling error.

mansinahar commented on stored_requests/backends/file_fetcher/fetcher.go:
```diff
@@ -49,7 +50,12 @@ func (fetcher *eagerFetcher) FetchAccount(ctx context.Context, accountID string)
 	}
-	return accountJSON, nil
+
+	completeJSON, err := jsonpatch.MergePatch(accountDefaultsJSON, accountJSON)
```
We should add test cases to cover this new change where default account information is provided to `FetchAccount` and it is properly merged and sent back.

Pubmatic-Dhruv-Sonone replied:
Added new test cases.

mansinahar commented on stored_requests/fetcher_test.go:
```diff
@@ -243,7 +243,7 @@ func TestAccountCacheHit(t *testing.T) {
 
 	metricsEngine.On("RecordAccountCacheResult", metrics.CacheHit, 1)
-	account, errs := aFetcherWithCache.FetchAccount(ctx, "known")
+	account, errs := aFetcherWithCache.FetchAccount(ctx, json.RawMessage("{}"), "known")
```
Might be a good idea to add another test case with some default account JSON

Pubmatic-Dhruv-Sonone replied:
This test cases is for account cache. The flow does not reach the account json merge code, as account fetcher is mocked.

mansinahar commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -62,10 +62,6 @@ func TestJsonSampleRequests(t *testing.T) {
 			"Makes sure we handle (default) aliased bidders properly",
 			"aliased",
 		},
-		{
```
May I ask what's the reason behind removing the malformed error test cases?

Pubmatic-Dhruv-Sonone replied:
Reverted

mansinahar reviewed (commented):
@Pubmatic-Dhruv-Sonone Thanks for making the changes. I have a couple more minor comments but other than that it looks great.

VeronikaSolovei9 reviewed (approved):
LGTM
